ui/components/component-library/button-base/button-base.js

```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+
+import Box from '../../ui/box';
+import { Icon, ICON_NAMES } from '../icon';
+import { Text } from '../text';
+
+import {
+  ALIGN_ITEMS,
+  JUSTIFY_CONTENT,
+  TEXT_COLORS,
+  TEXT,
+  SIZES,
+  FLEX_DIRECTION,
+} from '../../../helpers/constants/design-system';
+import { BUTTON_SIZES } from './button.constants';
+
+export const ButtonBase = ({
+  as = 'button',
+  block,
+  children,
+  className,
+  size = BUTTON_SIZES.MD,
+  icon,
+  iconPositionRight,
+  loading,
+  disabled,
+  ...props
+}) => {
+  return (
+    <Box
+      as={as}
+      paddingLeft={size === BUTTON_SIZES.AUTO ? 0 : 4}
+      paddingRight={size === BUTTON_SIZES.AUTO ? 0 : 4}
+      className={classnames(
+        'mm-button',
+        `mm-button--size-${size}`,
+        {
+          [`mm-button--loading`]: Boolean(loading),
+          [`mm-button--disabled`]: Boolean(disabled),
+          [`mm-button--block`]: Boolean(block),
+        },
+        className,
+      )}
+      disabled={disabled}
+      {...props}
+    >
+      <Text
+        as="span"
+        className="mm-button--content"
+        alignItems={ALIGN_ITEMS.CENTER}
+        justifyContent={JUSTIFY_CONTENT.CENTER}
+        flexDirection={
+          iconPositionRight ? FLEX_DIRECTION.ROW_REVERSE : FLEX_DIRECTION.ROW
+        }
+        gap={2}
+        variant={size === BUTTON_SIZES.AUTO ? TEXT.INHERIT : TEXT.BODY_MD}
+        color={TEXT_COLORS.INHERIT}
+      >
+        {icon && (
+          <Icon
+            name={icon}
+            size={size === BUTTON_SIZES.AUTO ? SIZES.AUTO : SIZES.SM}
+          />
```

Comment:
Just thought of this and it's likely a pattern that we will want to adopt moving forward. The ability to pass props to the nested components could be useful e.g. data test ids, aria-lablels, changing the nested components props for a custom component etc. Not required for this PR though.

```suggestion
          <Icon
            name={icon}
            size={size === BUTTON_SIZES.AUTO ? SIZES.AUTO : SIZES.SM}
            {...iconProps}
          />
```

Reply:
Make sure to add this to PropTypes and give it a description. I've also realized that I stuffed up the PropTypes in the Icon component. You can fix it by changing the `style: PropTypes.style` to `style: PropTypes.object` 

<img width="780" alt="Screen Shot 2022-09-28 at 12 54 00 PM" src="https://user-images.githubusercontent.com/8112138/192876292-a2552c63-845d-4522-bd1a-979aa6b86c4b.png">